geomstats/learning/frechet_mean.py

```
@@ -0,0 +1,348 @@
+            last_iteration, var, sq_dist))
+
+    mean = gs.to_ndarray(mean, to_ndim=2)
+    return mean
+
+
+def _ball_gradient_descent(points, metric, weights, n_max_iterations):
+    lr = 1e-3
+    tau = 5e-3
+
+    if len(points) == 1:
+        return points
+
+    iteration = 0
+    convergence = math.inf
+    barycenter = points.mean(0, keepdims=True) * 0
+
+    while convergence > tau and n_max_iterations > iteration:
+
+        iteration += 1
+
+        expand_barycenter = gs.repeat(barycenter, points.shape[0], 0)
+
+        grad_tangent = 2 * metric.log(points, expand_barycenter)
+
+        cc_barycenter = metric.exp(
+            lr * grad_tangent.sum(0, keepdims=True), barycenter)
+
+        convergence = metric.dist(cc_barycenter, barycenter).max().item()
+
+        barycenter = cc_barycenter
+
+    if iteration == n_max_iterations:
+        warnings.warn(
+            'Maximum number of iterations {} reached. The '
+            'mean may be inaccurate'.format(n_max_iterations))
+
+    return barycenter
+
+
+def _adaptive_gradient_descent(points,
+                               metric,
+                               weights=None,
+                               n_max_iterations=32,
+                               epsilon=1e-12,
+                               init_points=[]):
+    """Compute the Frechet mean using gradient descent.
+
+    Frechet mean of (weighted) points using adaptive time-steps
+    The loss function optimized is ||M_1(x)||_x (where M_1(x) is
+    the tangent mean at x) rather than the mean-square-distance (MSD)
+    because this saves computation time.
+
+    Parameters
+    ----------
+    points: array-like, shape=[n_samples, dimension]
+
+    weights: array-like, shape=[n_samples, 1], optional
+
+    init_points: array-like, shape=[n_init, dimension]
+
+    epsilon: tolerance for stopping the gradient descent
+    """
+    # TODO(xpennec): This function assumes that all points are lists
+    #  of vectors and not of matrices
```

maybe raise an error if `point_type=='matrix'`

---

Yes, done!